Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Нам даны следующие задания:

1.Посчитайте, сколько книг вышло после 1 января 2000 года;

2.Для каждой книги посчитайте количество обзоров и среднюю оценку;

3.Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

4.Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

5.Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

Для выполнения заданий мы

- опишим цели исследования;

- исследуем таблицы — выведите первые строки;

- сделаем по одному SQL-запросу для решения каждого задания;

- выведем результаты каждого запроса в тетрадке;

- опишем выводы по каждой из решённых задач.


Основной целью является успешное выполнение этого задания и получить финальный документ, подтверждающий прохождение обучения (йййййёёёёёё-хуууууу).

Для этого нам необходимо успешно инпортировать библиотеки и исследовать предоставленные нам таблицы.


<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Здорово, что не забываешь про оформление проекта
</div>

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [2]:
# изучим первые 5 строк таблицы books
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Таблица books

Содержит данные о книгах:

book_id — идентификатор книги;

author_id — идентификатор автора;

title — название книги;

num_pages — количество страниц;

publication_date — дата публикации книги;

publisher_id — идентификатор издателя.

In [3]:
# изучим первые 5 строк таблицы authors
query = '''SELECT * FROM authors LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Таблица authors

Содержит данные об авторах:

author_id — идентификатор автора;

author — имя автора.

In [4]:
# изучим первые 5 строк таблицы publishers
query = '''SELECT * FROM publishers LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Таблица publishers

Содержит данные об издательствах:

publisher_id — идентификатор издательства;

publisher — название издательства;

In [5]:
# изучим первые 5 строк таблицы ratings
query = '''SELECT * FROM ratings LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Таблица ratings

Содержит данные о пользовательских оценках книг:

rating_id — идентификатор оценки;

book_id — идентификатор книги;

username — имя пользователя, оставившего оценку;

rating — оценка книги.

In [6]:
# изучим первые 5 строк таблицы reviews
query = '''SELECT * FROM reviews LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Таблица reviews

Содержит данные о пользовательских обзорах:

review_id — идентификатор обзора;

book_id — идентификатор книги;

username — имя автора обзора;

text — текст обзора.

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Отлично первый взгляд на данные есть. Ты корректно выгрузила данные и посмотрела их содержимое
</div>

In [7]:
# Задание 1: посчитайте, сколько книг вышло после 1 января 2000 года
query = '''SELECT COUNT(book_id) FROM books WHERE publication_date > '2000-01-01' '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

После 1 января 2000 года было выпущено 819 книг.

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Расчёт верный
</div>

In [9]:
# Задание 2: для каждой книги посчитайте количество обзоров и среднюю оценку
query = '''SELECT title, SUM(count_reviews) AS count_reviews, AVG(avg_rating) AS avg_rating
                   FROM books 
                   LEFT JOIN (SELECT book_id, COUNT(review_id) AS count_reviews
                        FROM reviews                        
                        GROUP BY book_id) AS count_reviews
                   ON books.book_id = count_reviews.book_id
                   LEFT JOIN (SELECT book_id, AVG(rating) AS avg_rating
                        FROM ratings 
                        GROUP BY book_id) AS avg_rating
                   ON books.book_id = avg_rating.book_id
                   GROUP BY books.book_id                         
                                                         
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

title  count_reviews  \
0            The Body in the Library (Miss Marple  #3)            2.0   
1                                            Galápagos            2.0   
2                             A Tree Grows in Brooklyn            5.0   
3    Undaunted Courage: The Pioneering First Missio...            2.0   
4                    American Gods (American Gods  #1)            5.0   
..                                                 ...            ...   
995                                         Possession            3.0   
996          A Woman of Substance (Emma Harte Saga #1)            2.0   
997                                          Christine            3.0   
998  The Magicians' Guild (Black Magician Trilogy  #1)            2.0   
999                           The Plot Against America            2.0   

     avg_rating  
0      4.500000  
1      4.500000  
2      4.250000  
3      4.000000  
4      4.000000  
..          ...  
995    4.000000  
996    5.000000  
997    3.428571  
998    3.500000  
999    3.000000  

[1000 rows x 3 columns]

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №2__

Да, теперь всё корректно
</div>

Для каждой книги посчитано количество обзоров и средняя оценка.

In [10]:
# Задание 3: определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры
query = '''SELECT publisher
                          FROM books LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
                          WHERE num_pages > 50
                          GROUP BY publisher
                          ORDER BY COUNT(book_id) DESC                                                 
                                                         
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher
0                                        Penguin Books
1                                              Vintage
2                             Grand Central Publishing
3                                     Penguin Classics
4                                     Ballantine Books
..                                                 ...
329                                         Turtleback
330  Atheneum Books for Young Readers: Richard Jack...
331                                     Penguin Signet
332                                    Victor Gollancz
333                      Harvard Business Review Press

[334 rows x 1 columns]

Издательство, которое выпустило наибольшее число книг толще 50 страниц, - Penguin Books 

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Согласен с результатом. Действительно данное издательство является лидером по выпущенным книгам
</div>

In [12]:
# Задание 4: определим автора с самой высокой средней оценкой книг — учитывать будем только книги с 50 и более оценками
query = '''
SELECT a.author, AVG(rating)
FROM authors AS a
JOIN books AS b ON a.author_id = b.author_id
JOIN ratings AS r ON b.book_id = r.book_id
WHERE b.book_id IN (
    SELECT book_id
    FROM ratings 
    GROUP BY book_id
    HAVING COUNT(rating_id) >= 50
    ORDER BY COUNT(rating_id) DESC)
GROUP BY author
ORDER BY avg DESC                                          
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author       avg
0                          J.K. Rowling/Mary GrandPré  4.287097
1                   Markus Zusak/Cao Xuân Việt Khương  4.264151
2                                      J.R.R. Tolkien  4.246914
3                                   Louisa May Alcott  4.192308
4                                        Rick Riordan  4.080645
5                                     William Golding  3.901408
6                                       J.D. Salinger  3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
9                                          Lois Lowry  3.750000
10                                          Dan Brown  3.741259
11            George Orwell/Boris Grabnar/Peter Škerl  3.729730
12                                    Stephenie Meyer  3.662500
13                                     John Steinbeck  3.622951

Автором с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками, является J.K. Rowling/Mary GrandPré.

In [13]:
# Задание 5: посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок
query = '''SELECT AVG(count)
      FROM (SELECT COUNT(review_id)
            FROM reviews AS re
            LEFT JOIN (SELECT username, 
                              COUNT(rating_id)
                       FROM ratings
                       GROUP BY username) AS ra ON ra.username = re.username
            WHERE count > 48
            GROUP BY re.username) AS count                            
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №1__

Ты корректно отобрала пользователей и рассчитала их показатели.

Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок, = 24 обзора.

# ИТАК 

для счастливого достижениям пославленной нами цели, мы нашли ответы на все задания:

- задание 1 - 819 книг было выпущено после 01.01;

- задание 2 - для каждой книги посчитано количество обзоров и средняя оценка;

- задание 3 - издательство, которое выпустило наибольшее число книг толще 50 страниц, - Penguin Books;

- задание 4 - автором с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками, является J.K. Rowling/Mary GrandPré;

- задание 5 - в среднем 24 обзора оставляли пользователи, которые ставили больше 48 оценок.

<div class="alert alert-block alert-success">✔️
    

__Комментарий от тимлида №2__

Получилась хорошая работа, развернутый итоговый вывод придает ей вид законченного проекта. Поздравляю с окончанием курса! У тебя получилось справиться со всеми задачами. 
</div>